base/expr.jl

```diff
@@ -371,6 +371,187 @@ macro constprop(setting, ex)
+
+`@assume_effects` overrides the compiler's effect modeling for the given method.
+`ex` must be a method definition or `@ccall` expression.
+
+WARNING: Improper use of this macro causes undefined behavior (including crashes,
+incorrect answers, or other hard to track bugs). Use with care and only if absolutely
+required.
+
+In general, each `setting` value makes an assertion about the behavior of the
+function, without requiring the compiler to prove that this behavior is indeed
+true. These assertions are made for all world ages. It is thus advisable to limit
+the use of generic functions that may later be extended to invalidate the
+assumption (which would cause undefined behavior).
+
+The following `setting`s are supported.
+- `:consistent`
+- `:effect_free`
+- `:nothrow`
+- `:terminates_globally`
+- `:terminates_locally`
+- `:total`
+
+---
+# `:consistent`
+
+The `:consistent` setting asserts that for egal inputs:
+- The manner of termination (return value, exception, non-termination) will always be the same.
+- If the method returns, the results will always be egal.
+
+!!! note
+    This in particular implies that the return value of the method must be
+    immutable. Multiple allocations of mutable objects (even with identical
+    contents) are not egal.
+
+!!! note
+    The `:consistent`-cy assertion is made world-age wise. More formally, write
+    ``fᵢ`` for the evaluation of ``f`` in world-age ``i``, then we require:
+    ```math
+    ∀ i, x, y: x ≡ y → fᵢ(x) ≡ fᵢ(y)
+    ```
+    However, for two world ages ``i``, ``j`` s.t. ``i ≠ j``, we may have ``fᵢ(x) ≢ fⱼ(y)``.
+
+    A further implication is that `:consistent` functions may not make their
+    return value dependent on the state of the heap or any other global state
+    that is not constant for a given world age.
+
+!!! note
+    The `:consistent`-cy includes all legal rewrites performed by the optimizer.
+    For example, floating-point fastmath operations are not considered `:consistent`,
+    because the optimizer may rewrite them causing the output to not be `:consistent`,
+    even for the same world age (e.g. because one ran in the interpreter, while
+    the other was optimized).
+
+!!! note
+    If `:consistent` functions terminate by throwing an exception, that exception
+    itself is not required to meet the egality requirement specified above.
+
+---
+# `:effect_free`
+
+The `:effect_free` setting asserts that the method is free of externally semantically
+visible side effects. The following is an incomplete list of externally semantically
+visible side effects:
+- Changing the value of a global variable.
+- Mutating the heap (e.g. an array or mutable value), except as noted below
+- Changing the method table (e.g. through calls to eval)
+- File/Network/etc. I/O
+- Task switching
+
+However, the following are explicitly not semantically visible, even if they
+may be observable:
+- Memory allocations (both mutable and immutable)
+- Elapsed time
```

Comment:
What about:
```
while true; end
```
which is a method that always throws.

Reply:
That's what `terminates` is for.